src/domain/login/LoginViewModel.js

```diff
@@ -0,0 +1,187 @@
+            this._hideHomeserver = true;
+            this._completeSSOLoginViewModel = this.track(new CompleteSSOLoginViewModel(
+                this.childOptions(
+                    {
+                        sessionContainer: this._sessionContainer,
+                        attemptLogin: loginMethod => this.attemptLogin(loginMethod),
+                        showError: message => this.showError(message),
+                        loginToken: this._loginToken
+                    })));
+            this.emitChange("completeSSOLoginViewModel");
+        }
+        else {
+            this._errorMessage = "";
+            try {
+                this._loginOptions = await this._sessionContainer.queryLogin(homeserver);
+            }
+            catch (e) {
+                this._loginOptions = null;
+            }
+            if (this._loginOptions) {
+                if (this._loginOptions.sso) { this._showSSOLogin(); }
+                if (this._loginOptions.password) { this._showPasswordLogin(); }
+                if (!this._loginOptions.sso && !this._loginOptions.password) {
+                    this.showError("This homeserver neither supports SSO nor Password based login flows");
+                } 
+            }
+            else {
+                this.showError("Could not query login methods supported by the homeserver");
+            }
+        }
+    }
+
+    _showPasswordLogin() {
+        this._passwordLoginViewModel = this.track(new PasswordLoginViewModel(
+            this.childOptions({
+                sessionContainer: this._sessionContainer,
+                loginOptions: this._loginOptions,
+                homeserver: this._homeserver,
+                attemptLogin: loginMethod => this.attemptLogin(loginMethod),
+                showError: message => this.showError(message)
+        })));
+        this.emitChange("passwordLoginViewModel");
+    }
+
+    _showSSOLogin() {
+        this._startSSOLoginViewModel = this.track(
+            new StartSSOLoginViewModel(this.childOptions({loginOptions: this._loginOptions}))
+        );
+        this.emitChange("startSSOLoginViewModel");
+    }
+
+    showError(message) {
+        this._errorMessage = message;
+        this.emitChange("errorMessage");
+        this._errorMessage = "";
+    }
+
+    _toggleBusy(status) {
+        this._isBusy = status;
+        this.emitChange("isBusy");
+    }
+
+    async attemptLogin(loginMethod) {
+        this._toggleBusy(true);
+        this._sessionContainer.startWithLogin(loginMethod);
+        const loadStatus = this._sessionContainer.loadStatus;
+        const handle = loadStatus.waitFor(status => status !== LoadStatus.Login);
+        await handle.promise;
+        this._toggleBusy(false);
+        const status = loadStatus.get();
+        if (status === LoadStatus.LoginFailed) {
+            return this._sessionContainer.loginFailure;
+        }
+        this._hideHomeserver = true;
+        this._disposeViewModels();
```

Comment:
Is it okay to dispose the viewmodels when said viewmodels are awaiting on this method?

Reply:
Hmm, yeah, it's not ideal but as the callee should not do anything if there is no error, I think it's probably fine in practice. I also don't see a better way.